reference-implementation/include/emitc/types.h

```diff
@@ -88,25 +88,104 @@ struct switch_t<case_t<B, T>> {
   static_assert(B, "None of the supplied conditions evaluate to true.");
   using type = T;
 };
+
+// Wrapper class to prevent use of the template specialization for
+// std::vector<bool> in the Tensor class
+template <typename T>
+class TypeWrapper {
+public:
+  TypeWrapper() = default;
+
+  TypeWrapper(const T &value) : value(value) {}
+
+  TypeWrapper<T> &operator=(const T &value) {
+    this->value = value;
+    return *this;
+  }
+
+  operator T &() { return value; }
+
+  constexpr operator const T &() const { return value; }
+
+private:
+  T value;
+};
+
+template <typename T, typename Iterator>
+class IteratorWrapper {
+public:
+  using iterator_category = std::forward_iterator_tag;
+  using value_type = T;
+  using difference_type = std::ptrdiff_t;
+  using pointer = T *;
+  using reference = T &;
+
+  IteratorWrapper() = default;
+
+  IteratorWrapper(const Iterator &it) : iterator(it) {}
+
+  reference operator*() const { return reinterpret_cast<reference>(*iterator); }
+
+  IteratorWrapper<T, Iterator> &operator++() {
+    ++iterator;
+    return *this;
+  }
+
+  IteratorWrapper<T, Iterator> operator++(int) {
+    IteratorWrapper<T, Iterator> result = *this;
+    ++iterator;
+    return result;
+  }
+
+  IteratorWrapper<T, Iterator> &operator+=(difference_type d) {
+    iterator += d;
+    return *this;
+  }
+
+  IteratorWrapper<T, Iterator> operator+(difference_type d) const {
+    return IteratorWrapper<T, Iterator>(iterator + d);
+  }
+
+  bool operator==(const IteratorWrapper &other) const {
+    return iterator == other.iterator;
+  }
+
+  bool operator!=(const IteratorWrapper &other) const {
+    return iterator != other.iterator;
+  }
+
+private:
+  Iterator iterator;
+};
 } // namespace detail
 
 /// The elements of a Tensor are stored contiguously in memory in a
 /// row-major layout.
 template <typename T, size_t... Shape>
 class Tensor {
+private:
+  using storage_type = detail::TypeWrapper<T>;
+
 public:
   using value_type = T;
-  using reference = typename std::vector<T>::reference;
-  using iterator = typename std::vector<T>::iterator;
-  using const_iterator = typename std::vector<T>::const_iterator;
+  static_assert(sizeof(T) == sizeof(storage_type),
+                "storage_type has a different size than the raw type T");
+  static_assert(std::is_standard_layout_v<storage_type>,
```

Comment:
```suggestion
  static_assert(std::is_standard_layout<storage_type>::value,
```
I don't think we have the defined the language standard of the generated code anywhere.

Reply:
I don't think we've set the C++ standard, but C++14 is reasonable. Depending on the code we generate in the conversion paths it is possible to require C++20 (see https://mlir.llvm.org/docs/Dialects/EmitC/), but without further checking I don't know what our reference implementation requires so far.

Reply:
Since this change just resolves syntactic sugar I will still apply it. Also because it might fix the failed workflow run.